plugins/genesys/fu-genesys-hubhid-device.c

```diff
@@ -0,0 +1,400 @@
+/*
+ * Copyright (C) 2023 Adam.Chen <Adam.Chen@genesyslogic.com.tw>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <string.h>
+
+#include "fu-genesys-common.h"
+#include "fu-genesys-hubhid-device.h"
+
+#define GENESYS_HUBHID_REPORT_ID	  0
+#define GENESYS_HUBHID_REPORT_BYTE_LENGTH 0x40
+#define GENESYS_HUBHID_REPORT_TIMEOUT	  100 /* ms */
+#define GENESYS_HUBHID_REPORT_FLAGS	  FU_HID_DEVICE_FLAG_ALLOW_TRUNC | FU_HID_DEVICE_FLAG_IS_FEATURE
+
+typedef union {
+	struct {
+		guint8 recipient : 2; /* GUsbDeviceRecipient */
+		guint8 reserved : 3;
+		guint8 type : 2; /* GUsbDeviceRequestType */
+		guint8 dir : 1;	 /* GUsbDeviceDirection */
+	};
+	guint8 bm;
+} FuGenesysUsbRequestType;
+
+typedef struct {
+	FuGenesysUsbRequestType req_type;
+	guint8 request;
+	guint16 value;
+	guint16 index;
+	guint16 length;
+} FuGenesysUsbSetup;
+
+struct _FuGenesysHubhidDevice {
+	FuHidDevice parent_instance;
+
+	gboolean support_report_pack;
+	guint16 report_length;
+	guint16 max_report_pack_data_length;
+};
+
+G_DEFINE_TYPE(FuGenesysHubhidDevice, fu_genesys_hubhid_device, FU_TYPE_HID_DEVICE)
+
+static gboolean
+fu_genesys_hubhid_device_command_read(FuGenesysHubhidDevice *self,
+				      FuGenesysUsbSetup *setup,
+				      guint8 *data,
```

Comment:
i think include `gsize datasz` and then `g_return_val_if_fail(datasz == setup->length, FALSE)`